plugins/storage/rdbms/list_jobs.go

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+//
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+package rdbms
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/facebookincubator/contest/pkg/job"
+	"github.com/facebookincubator/contest/pkg/storage"
+	"github.com/facebookincubator/contest/pkg/types"
+)
+
+func (r *RDBMS) ListJobs(query *storage.JobQuery) ([]types.JobID, error) {
+	res := []types.JobID{}
+
+	// Quoting SQL strings is hard. https://github.com/golang/go/issues/18478
+	// For now, just disallow anything that is not [a-zA-Z0-9_-]
+	if err := job.CheckTags(query.Tags); err != nil {
+		return nil, err
+	}
+
+	// Job state is updated by framework events, ensure there aren't any pending.
+	r.frameworkEventsLock.Lock()
+	err := r.flushFrameworkEvents()
+	r.frameworkEventsLock.Unlock()
+	if err != nil {
+		return nil, fmt.Errorf("could not flush events before reading events: %v", err)
+	}
+
+	// Construct the query.
+	parts := []string{"SELECT jobs.job_id FROM jobs"}
+	// Tag filtering uses joins, 1 per tag.
+	for i := range query.Tags {
+		parts = append(parts, fmt.Sprintf(`INNER JOIN job_tags jt%d ON jobs.job_id = jt%d.job_id`, i, i))
```

Comment:
This is super super scary. We are one innocent regex change away from a full blown SQL injection in a read-only API call that does not even have ACLs of its own.

Would it maybe be possible to change this to append static strings `INNER JOIN job_tags ? ON jobs.job_id = ?` and keep a `[]interface{}` with the arguments in the same order to later pass it to `db.Query`? (Not entirely sure, not everything can be used with `?` (also discussed in the ticket you linked above).

Reply:
oh, i totally agree and i went looking for a quoting function to do exactly what you are suggesting.
unfortunately, Go does not provide such interface in sql or sql driver interface, this is what https://github.com/golang/go/issues/18478 is about.
moreover, in general, it is not even possible to do it for Mysql with a simple function, as quoting converntion depends on connection settings! in particular, whether backslash escape sequences are interpreted or not. it's a pain.
so, in the end i settled for a restrictive regex as a means of ensuring there is no injection vector.

Reply:
Yeah I get the part about the quoting functions. I was just wondering if it was possible to construct the query with something like `INNER JOIN job_tags ? ON jobs.job_id = ?` and then pass the parameters into `db.Exec()`, as those are always properly escaped, automatically.

Reply:
oh, yes, you are right. done

Reply:
Did this change, I don't see it. Or maybe I misunderstood?

Reply:
it did. there is still some Sprintf'ing happening there, but it's only for counters, all user-supplied data is passed via the `qargs` slice and substituted by the driver.